Feat: add sap-description-column-label ESLint rule (#4542)
SAP/open-ux-tools: packages/eslint-plugin-fiori-tools/docs/rules/sap-description-column-label.md

# Require Meaningful Labels for Description (Text) Properties (`sap-description-column-label`)

> Ensures that the property referenced as the description (text) value via the `Common.Text` annotation has a meaningful `Common.Label` annotation — not a generic value such as `"Name"` or `"Description"`, and not the same label as the ID property it describes.

Reviewer: ```suggestion
Ensures that the property referenced as the description (text) value using the `Common.Text` annotation has a meaningful `Common.Label` annotation. The label must not be a generic value such as `"Name"` or `"Description"` nor the same label as the ID property it describes.
```

Author: Done

## Rule Details

> When a field uses the `Common.Text` annotation to associate a human-readable description property with a technical key property, the description property should carry a `Common.Label` annotation that clearly identifies its content. A generic label like `"Name"` or `"Description"` provides no useful context to the user in the UI (for example, as a column header in a table), and a label that is identical to the ID property's label creates ambiguity between the two columns.

Reviewer: ```suggestion
When a field uses the `Common.Text` annotation to associate a human-readable description property with a technical key property, the description property must carry a `Common.Label` annotation that clearly identifies its content. A generic label such as `"Name"` or `"Description"` provides no useful context to the user in the UI and a label that is identical to the ID property's label creates ambiguity.
```

Author: Done

> The rule checks every `Common.Text` annotation on a property that belongs to an entity type used in the application (i.e. linked to at least one page via the manifest). For each such annotation the rule:

Reviewer: ```suggestion
The rule checks every `Common.Text` annotation on a property that belongs to an entity type used in the application, that is linked to at least one page in the `manifest.json` file. For each annotation, the rule performs the following:
```

Author: Done

> 1. Resolves the path to the referenced text property (including navigation segments such as `category/name`).

Reviewer: ```suggestion
1. Resolves the path to the referenced text property which includes navigation segments such as `category/name`.
```

Author: Done

2. Reads the `Common.Label` annotation on that property.

> 3. Flags the annotation if the label is `"Name"` or `"Description"` (case-insensitive) — **trivialLabel**.

Reviewer: ```suggestion
3. Produces a `trivialLabel` warning if the label is the same as the `"Name"` or `"Description"` (case-insensitive).
```

Author: Done

> 4. Flags the annotation if the label is identical (case-insensitive) to the `Common.Label` annotation of the ID property that carries the `Common.Text` annotation — **duplicateLabel**.

Reviewer: ```suggestion
4. Produces a `duplicateLabel` warning if the label is identical (case-insensitive) to the `Common.Label` annotation of the ID property that carries the `Common.Text` annotation.
```

Author: Done

> If the text property has no `Common.Label` annotation, the rule does not flag it.

Reviewer: ```suggestion
If the text property has no `Common.Label` annotation, the rule does not produce a warning.
```

Author: Done

> > **OData V2**: The rule also applies to OData V2 services. In V2 metadata, the `sap:text` attribute on a `Property` element is treated as an implicit `Common.Text` annotation, and the `sap:label` attribute is treated as an implicit `Common.Label` annotation. The rule reports on the metadata file directly when these inline attributes produce a trivial or duplicate label.

Reviewer: ```suggestion
> **OData V2**: This rule also applies to OData V2 services. In OData V2 metadata, the `sap:text` attribute on a `Property` element is treated as an implicit `Common.Text` annotation and the `sap:label` attribute is treated as an implicit `Common.Label` annotation. The rule reports on the metadata file directly when these inline attributes produce a trivial or duplicate label.
```

Author: Done

### Why Was This Rule Introduced?

> When a property is displayed using a description column (e.g. `TextArrangement/TextOnly`), the column header comes from the `Common.Label` annotation of the text property. A label of `"Name"` or `"Description"` is semantically meaningless in that context — it does not tell the user _which_ name or description is shown. Similarly, if the description column carries the exact same label as the ID column, the user cannot distinguish the two columns.

Reviewer: ```suggestion
When a property is displayed using a description column, for example, `TextArrangement/TextOnly`, the column header comes from the `Common.Label` annotation of the text property. A label of `"Name"` or `"Description"` is semantically meaningless in that context, it does not tell the user which name or description is shown. Similarly, if the description column carries the exact same label as the ID column, the user cannot distinguish the two columns.
```

Author: Done

### Warning Messages

> **trivialLabel**

Reviewer: ```suggestion
`trivialLabel`
```

Author: Done

```

> The text property "{{textPropertyTarget}}" has a generic label "{{textPropertyLabel}}". Use a more descriptive label that distinguishes it from other properties.

Reviewer: ```suggestion
The "{{textPropertyTarget}}" text property has a "{{textPropertyLabel}}" generic label. Use a more descriptive label that distinguishes it from other properties.
```

Author: Done

```

> **duplicateLabel**

Reviewer: ```suggestion
`duplicateLabel`
```

Author: Done

```

> The text property "{{textPropertyTarget}}" has the same label "{{textPropertyLabel}}" as the ID property "{{idPropertyTarget}}". The description column label should be different from the ID label.

Reviewer: ```suggestion
The "{{textPropertyTarget}}" text property has the same "{{textPropertyLabel}}" label as the "{{idPropertyTarget}}" ID property. The description column label must be different from the ID label.
```

Author: Done

```

## Examples

### Incorrect Annotations

> **`trivialLabel` — the `Common.Label` annotation on the text property is too generic:**

Reviewer: ```suggestion
`trivialLabel`: The `Common.Label` annotation on the text property is too generic:
```

Author: Done

```xml

> <!-- ProductBaseUnit uses UnitOfMeasure_Text as its description column via Common.Text annotation -->

Reviewer: ```suggestion
<!-- ProductBaseUnit uses UnitOfMeasure_Text as its description column using the Common.Text annotation -->
```

Author: Done with commit 49af900

<Annotations Target="MyService.MyEntity/ProductBaseUnit">
    <Annotation Term="Common.Text" Path="to_BaseUnit/UnitOfMeasure_Text"/>
</Annotations>

> <!-- The Common.Label annotation on the description property is "Name" — too generic -->

Reviewer: ```suggestion
<!-- The Common.Label annotation on the description property is "Name" which is too generic -->
```

Author: Done

<Annotations Target="MyService.BaseUnitType/UnitOfMeasure_Text">
    <Annotation Term="Common.Label" String="Name"/>
</Annotations>
```

> **`duplicateLabel` — the `Common.Label` annotation on the text property is identical to the ID property label:**

Reviewer: ```suggestion
`duplicateLabel`: The `Common.Label` annotation on the text property is identical to the ID property label:**
```

Author: Done (with trailing ** removed)

```xml

> <!-- Supplier uses CompanyName as its description column via Common.Text annotation, both properties have label "Supplier" -->

Reviewer: ```suggestion
<!-- Supplier uses CompanyName as its description column using the Common.Text annotation but both properties have the "Supplier" label -->
```

Author: Done

<Annotations Target="MyService.MyEntity/Supplier">
    <Annotation Term="Common.Text" Path="to_Supplier/CompanyName"/>
    <Annotation Term="Common.Label" String="Supplier"/>
</Annotations>

> <!-- Common.Label annotation identical to the ID property above — ambiguous column headers -->

Reviewer: ```suggestion
<!-- Common.Label annotation identical to the ID property above which leads to ambiguous column headers -->
```

Author: Done

<Annotations Target="MyService.SupplierType/CompanyName">
    <Annotation Term="Common.Label" String="Supplier"/>
</Annotations>
```

> **OData V2 `trivialLabel` — inline `sap:label` on the text property is too generic (reported on the metadata file):**

Reviewer: ```suggestion
**trivialLabel` for OData V2: The inline `sap:label` on the text property is too generic. This is reported on the metadata file:
```

Author: Done

```xml
<!-- sap:text attribute acts as an implicit Common.Text annotation -->
<Property Name="Product" Type="Edm.String" sap:text="ProductName" sap:label="Product ID" />
<!-- sap:label="Name" acts as an implicit Common.Label annotation — too generic -->
<Property Name="ProductName" Type="Edm.String" sap:label="Name" />
```

> **OData V2 `duplicateLabel` — inline `sap:label` values on ID and text property are identical:**

Reviewer: ```suggestion
 `duplicateLabel` for OData V2: The inline `sap:label` values on ID and text property are identical:
```

Author: Done

```xml

> <!-- Both properties carry sap:label="Quantity Unit" via inline sap:label attribute -->

Reviewer: ```suggestion
<!-- Both properties carry sap:label="Quantity Unit" using the inline sap:label attribute -->
```

Author: Done

<Property Name="QuantityUnit" Type="Edm.String" sap:text="QuantityUnitT" sap:label="Quantity Unit" />
<Property Name="QuantityUnitT" Type="Edm.String" sap:label="Quantity Unit" />
```

### Correct Annotations

> **Meaningful, distinct `Common.Label` annotation on the text property:**

Reviewer: ```suggestion
The `Common.Label` annotation on the text property is meaningful and unique:
```

Author: Done

```xml
<Annotations Target="MyService.MyEntity/ProductBaseUnit">
    <Annotation Term="Common.Text" Path="to_BaseUnit/UnitOfMeasure_Text"/>
</Annotations>

<!-- Common.Label annotation with a descriptive value that distinguishes the column from the ID column -->
<Annotations Target="MyService.BaseUnitType/UnitOfMeasure_Text">
    <Annotation Term="Common.Label" String="Unit of Measure"/>
</Annotations>
```

> **Different `Common.Label` annotations on ID and text properties:**

Reviewer: ```suggestion
The `Common.Label` annotations on ID and text properties are unique:
```

Author: Done

```xml
<Annotations Target="MyService.MyEntity/Supplier">
    <Annotation Term="Common.Text" Path="to_Supplier/CompanyName"/>
    <Annotation Term="Common.Label" String="Supplier ID"/>
</Annotations>

<Annotations Target="MyService.SupplierType/CompanyName">
    <Annotation Term="Common.Label" String="Company Name"/>
</Annotations>
```

> **OData V2 — distinct `sap:label` values on ID and text property:**

Reviewer: ```suggestion
The `sap:label` values on ID and text property for OData V2 are unique:
```

Author: Done

```xml
<Property Name="QuantityUnit" Type="Edm.String" sap:text="QuantityUnitT" sap:label="Quantity Unit" />

> <!-- Distinct sap:label — unambiguous column headers -->

Reviewer: ```suggestion
<!-- The sap:label is unique which provides unambiguous column headers -->
```

Author: Done

<Property Name="QuantityUnitT" Type="Edm.String" sap:label="Unit of Measure Text" />
```

## Bug Report

If you encounter an issue with this rule, please open a [GitHub issue](https://github.com/SAP/open-ux-tools/issues).

## When Not to Disable This Rule

> This rule should not be disabled unless you have a confirmed backend constraint that prevents renaming the label. In most cases the fix is straightforward: update the `Common.Label` annotation (or the `sap:label` attribute in OData V2 metadata) on the text property to something meaningful and distinct from the ID property label.

Reviewer: ```suggestion
This rule must not be disabled unless you have a confirmed back-end constraint that prevents renaming the label. Update the `Common.Label` annotation or the `sap:label` attribute for OData V2 on the text property to something meaningful and distinct from the ID property label.
```

Author: Done
